### PR TITLE
Add Agentless Datadog Trace Exporter

### PR DESCRIPTION
**Pull request by ericmustin**
This is probably at a point where it could benefit from review, cc @mx-psi , some things to note:

- There's still some uncertainty around a few trace's span translation choices, as well as how to grab the global config stuff like hostname and env and apply it to the trace payloads (as well as the order of precedence for things like `DD_SERVICE` as opposed to the service already set by the tracer). but for the most part the implementation is complete, as is the flushing of trace and stats payloads.

- the most of what's in `/apm`, including tests and the snapshots, i've just grabbed [from serverless](https://github.com/DataDog/datadog-serverless-functions/tree/master/aws/logs_monitoring/trace_forwarder/internal/apm). I tried to remove everything that wasn't being used by our implementation, but didn't update the serverless tests, so they'll probably fail. 

- The big things here are going to be:
  - having a better set of eyes around my very hacky Go code. Anything you see that looks unusual, or could be optimised, there's likely _not_ a good reason i've chosen not to, it's just because i don't know Go especially well.
  - The implementation itself around converting Internal Traces to Datadog Traces largely follows the approach the [Zipkin Exporter](https://github.com/open-telemetry/opentelemetry-collector/tree/master/exporter/zipkinexporter) took. There are a few things zipkin is doing that we don't support and vice versa but the methods and code paths are similar.
  - The conventions of the coercing the internal spans to datadog spans largely follow the [python exporter's translation implementation](https://github.com/open-telemetry/opentelemetry-python/blob/b2559409b2bf82e693f3e68ed890dd7fd1fa8eae/exporter/opentelemetry-exporter-datadog/src/opentelemetry/exporter/datadog/exporter.py#L229), while the conventions and approach around generating+aggregating+flushing trace+stats payloads is largely based on the [serverless implementation](https://github.com/DataDog/datadog-serverless-functions/tree/master/aws/logs_monitoring/trace_forwarder/internal/apm). As i mentioned previously, I more or less ported everything in `/apm` directly from that repo.
  - I've been testing this locally using our `otel-collector-e2e-tests` package, lmk if you have any trouble with that.
  - How to appropriately manage the config. it's possible my approach has drifted from what we've ended up doing in the metric exporter.
  - Tests are going to be a hassle, If you see any shortcuts here or good similar test setups in other parts of the repo that I can use as a guide, let me know. I will try to start writing tests for this stuff next

### CODE REVIEW

File: exporter/datadogexporter/translate_traces.go
Comment (mx-psi) [inline exporter/datadogexporter/translate_traces.go]: I don't think we need the `msg` struct tags here: they are used for MessagePack which we are not using (right?). We can add them back in the future if we add support for sending traces to the Trace Agent.

File: exporter/datadogexporter/translate_traces.go
Comment (mx-psi) [inline exporter/datadogexporter/translate_traces.go]: Should we handle this during config sanitizing? I think it would make sense, since this should be handled for host tags too

File: exporter/datadogexporter/translate_traces.go
Comment (mx-psi) [inline exporter/datadogexporter/translate_traces.go]: I think we could simplify this function if we move this last bit outside of the function. That way we can use `traces` as named result parameter and avoid the `if resourceSpans.Len() == 0` bit.

File: exporter/datadogexporter/translate_traces.go
Comment (mx-psi) [inline exporter/datadogexporter/translate_traces.go]: Is this generally set by instrumentation libraries? And I am assuming this is set as an attribute, right?

File: exporter/datadogexporter/translate_traces.go
Comment (mx-psi) [inline exporter/datadogexporter/translate_traces.go]: nit: This `else` is unnecessary since we return on the `if` branch

File: exporter/datadogexporter/translate_traces.go
Comment (mx-psi) [inline exporter/datadogexporter/translate_traces.go]: I found this confusing; maybe add a comment here saying that this time is handled on `addToAPITrace`?

File: exporter/datadogexporter/translate_traces.go
Comment (mx-psi) [inline exporter/datadogexporter/translate_traces.go]:
```suggestion
	// it's possible end time is unset, so default to 0 rather than using a negative number
```

File: exporter/datadogexporter/translate_traces.go
Comment (mx-psi) [inline exporter/datadogexporter/translate_traces.go]: You should run `make fmt` to format the code

File: exporter/datadogexporter/translate_traces.go
Comment (mx-psi) [inline exporter/datadogexporter/translate_traces.go]: I think we can do
```suggestion
		duration = 0
```
can't we?

File: exporter/datadogexporter/translate_traces.go
Comment (mx-psi) [inline exporter/datadogexporter/translate_traces.go]: We should get the service from `TagsConfig.Service`.

File: exporter/datadogexporter/translate_traces.go
Comment (mx-psi) [inline exporter/datadogexporter/translate_traces.go]: nit: we can simplify this since `span.Error` is set to 0 by default (or leave it /a comment for emphasis)
```suggestion
		}
```

File: exporter/datadogexporter/translate_traces.go
Comment (mx-psi) [inline exporter/datadogexporter/translate_traces.go]: ditto
```suggestion
	}
```

**Review by mx-psi — COMMENTED** [review]
I reviewed some parts of the code and left a few comments so that you can keep working on this; most importantly I think you need to run `make fmt` and `make` to run all the linters and tests that can help catch errors (we will need them to pass to have a PR accepted upstream)

File: exporter/datadogexporter/translate_traces.go
Comment (ericmustin) [reply exporter/datadogexporter/translate_traces.go]: I believe it would be an attribute, probably on the resource? but i dont think anything is standard, we should probably defer to `env` or `environment` attribute on either the resource, instrumentation library, or span level, but going to try to see if there's anything in spec on this

Comment (ericmustin) [reply exporter/datadogexporter/translate_traces.go]: so i spoke with one of the tracing client maintainers, there's a non-required resource convention `deployment.environment` https://github.com/open-telemetry/opentelemetry-specification/blob/4b6f6e502e48e49b3bca93b13f309f9a61d5e1f7/specification/resource/semantic_conventions/deployment_environment.md#deployment that seems like the 'official otel' way of doing env. so let's defer to that, but if it doesn't exist, use DD_ENV / our env config

Comment (mx-psi) [reply exporter/datadogexporter/translate_traces.go]: That sounds good to me, marking this as solved

Comment (mx-psi) [reply exporter/datadogexporter/translate_traces.go]: Since this is solved we should remove the comment
```suggestion
```

File: exporter/datadogexporter/translate_traces.go
Comment (mx-psi) [inline exporter/datadogexporter/translate_traces.go]: nit: It makes more sense to pass the whole `TagsConfig` struct here since we will be using most if not all fields.

File: exporter/datadogexporter/traces_exporter.go
Comment (mx-psi) [inline exporter/datadogexporter/traces_exporter.go]: Can we get away with using OpenTelemetry retries here? If `pushTraceData` returns an error the exporter helper will deal with retrying. If we could avoid the `SendStats` and `SendTraces` using retries we could simplify this code a lot. 

The problem here would be if e.g. `SendStats` fails but `SendTraces` does not. We would have to signal failure here and so we would be sending traces twice. Would the endpoint be okay with that (and with the reversed case)?

File: exporter/datadogexporter/internal/apm/model.go
Comment (mx-psi) [inline exporter/datadogexporter/internal/apm/model.go]: We can make this private (lowercase) since it is only used in tests.

File: exporter/datadogexporter/internal/apm/model.go
Comment (mx-psi) [inline exporter/datadogexporter/internal/apm/model.go]: We probably don't need this, right? Although removing it will probably break tests of the serverless part :(

File: exporter/datadogexporter/internal/apm/trace-connection.go
Comment (mx-psi) [inline exporter/datadogexporter/internal/apm/trace-connection.go]: Name of the file should be in snake case, as in `trace_connection.go`

File: exporter/datadogexporter/internal/apm/model.go
Comment (mx-psi) [inline exporter/datadogexporter/internal/apm/model.go]: I am not sure if upstream reviewers would be okay with having our copyright headers here :/

File: exporter/datadogexporter/translate_traces.go
Comment (mx-psi) [inline exporter/datadogexporter/translate_traces.go]: Why do we assume an internal server error if the status code is not found in our map?

File: exporter/datadogexporter/translate_traces.go
Comment (mx-psi) [inline exporter/datadogexporter/translate_traces.go]: Since we know the `val`s are strings, wouldn't it make more sense to use `setStringTag` here? And then I am  not sure what do we have `setTag` for, is it for the commented code above?

**Review by mx-psi — COMMENTED** [review]
Second round; `impi` (after running `make fmt`) and `misspell` still give some errors that need to be addressed

File: exporter/datadogexporter/translate_traces.go
Comment (mx-psi) [reply exporter/datadogexporter/translate_traces.go]: I see this was removed but not added to `Config.Sanitize`, was this intended?

File: exporter/datadogexporter/translate_traces.go
Comment (ericmustin) [reply exporter/datadogexporter/translate_traces.go]: gotcha, makes sense will update

File: exporter/datadogexporter/traces_exporter.go
Comment (ericmustin) [reply exporter/datadogexporter/traces_exporter.go]: i believe duping 1/2 would be a possibility here, the endpoint is pretty naive iirc. There's a note in the serverless repo on this. 

Thinking about it though, it probably makes sense to prioritize `SendTraces` imo, since metrics are based on the traces, so if there's something unaccepted with the traces we probably don't want to trust the metrics that are generated anyway

File: exporter/datadogexporter/internal/apm/model.go
Comment (ericmustin) [reply exporter/datadogexporter/internal/apm/model.go]: I just went ahead and removed this along with the tests...i think we're gonna have to write these tests from scratch since we're dealing with a pretty significantly pared down version of the serverless code

File: exporter/datadogexporter/internal/apm/model.go
Comment (ericmustin) [reply exporter/datadogexporter/internal/apm/model.go]: yup, removed

File: exporter/datadogexporter/internal/apm/model.go
Comment (ericmustin) [reply exporter/datadogexporter/internal/apm/model.go]: whoops, updated

File: exporter/datadogexporter/translate_traces.go
Comment (ericmustin) [reply exporter/datadogexporter/translate_traces.go]: 🤷  just borrowing from the open census code (i think?) here. let me see if there's a more reasonable default

File: exporter/datadogexporter/translate_traces.go
Comment (ericmustin) [reply exporter/datadogexporter/translate_traces.go]: this is sorta a frankenstein left over from converting from open census...i'll try to clean this up

File: exporter/datadogexporter/translate_traces.go
Comment (ericmustin) [reply exporter/datadogexporter/translate_traces.go]: so looking into the collector a bit this seems to be what folks are using for "unknown error", so i think it's a reasonable default imo...we can see what they say in review.

File: exporter/datadogexporter/translate_traces.go
Comment (ericmustin) [reply exporter/datadogexporter/translate_traces.go]: updated this to just rely on setStringTag

File: exporter/datadogexporter/internal/apm/trace-connection.go
Comment (ericmustin) [reply exporter/datadogexporter/internal/apm/trace-connection.go]: 👍 updated

File: exporter/datadogexporter/translate_traces.go
Comment (mx-psi) [reply exporter/datadogexporter/translate_traces.go]: Okay, if that's what others are doing it makes sense to me!

**Comment by NathanBaulch** [comment]
Well that's annoying, I just stumbled on this branch/PR after successfully migrating the tracing portion of `opencensus-go-exporter-datadog` to OpenTelemetry [here](https://github.com/NathanBaulch/opencensus-go-exporter-datadog/commit/4024a9d69c078a1d567c548fcc540dda4f69e256). I admit it did seem weird that this work hadn't been done already (at least not somewhere easily discoverable).

Oh well, mine works but is just a minimal 1-1 (where possible) port, so I look forward to this official one being released. :+1:

While I'm here (and sorry if off-topic), I'm curious about the pros and cons of an OTEL exporter vs an OTEL API compatible wrapper as used in `dd-trace-go` ([example](https://github.com/DataDog/dd-trace-go/tree/v1/ddtrace/opentracer))?

Also, does "agentless" mean it'll go direct to datadoghq without dd-agent? Thanks.

File: exporter/datadogexporter/trace_connection.go
Comment (mx-psi) [inline exporter/datadogexporter/trace_connection.go]: Does this user agent make sense for Otel too? `xray-converter` seems to refer to AWS X-Ray

File: exporter/datadogexporter/trace_connection.go
Comment (mx-psi) [inline exporter/datadogexporter/trace_connection.go]: I know this code isn't yours but... if the error isn't retriable, shouldn't we return `nil` here? 

Thinking a bit about this I don't think it's necessarily the case that 4XX errors are not retriable e.g. if you are rate limited so I am even more confused by this

File: exporter/datadogexporter/trace_connection.go
Comment (mx-psi) [inline exporter/datadogexporter/trace_connection.go]: Not sure if we want to retry if we `err` here, if we can't build the request object it means something is wrong in our code (like, we used an invalid method or we had a malformed URL), I think we should log the error and return `nil`

**Comment by ericmustin** [comment]
@NathanBaulch the goal here is we really need to be able to live in the otel-collector. simply put, existing otel users likely aren't going to want to run a datadog agent along with an opentelemetry-collector (maybe they will? we plan to add agent export as well down the road). So, yes agentless means direct to api intake. We're still iterating here, as circumventing the agent and our own tracing clients introduces a whole can of worms around metric computation but we hope to have a PR in with the Otel folks by EOW. 

as far as `the pros and cons of an OTEL exporter vs an OTEL API compatible wrapper`, i think it's different use cases, an OTEL Exporter that lives in the opentelemetry-collector would be language agnostic but require the opentelemetry collector to be running, whereas the OTEL API Compatibility wrapper would leverage existing DD tracing architecture.

Sorry to hear about the duped work (nice work tho, fwiw!), but we can certainly keep you updated here on how this work is going and we'd love any feedback.

File: exporter/datadogexporter/stats.go
Comment (mx-psi) [inline exporter/datadogexporter/stats.go]: We should double check if this applies to Otel and update the comment

File: exporter/datadogexporter/traces_exporter.go
Comment (mx-psi) [inline exporter/datadogexporter/traces_exporter.go:85]: What pipeline does OTEL have?

File: exporter/datadogexporter/model.go
Comment (mx-psi) [inline exporter/datadogexporter/model.go]:
```suggestion
```

File: exporter/datadogexporter/model.go
Comment (mx-psi) [inline exporter/datadogexporter/model.go:80]: I am not sure what this function does, should we add a comment? Maybe it's obvious coming from an APM background but from an outsider's point of view it's unclear

File: exporter/datadogexporter/translate_traces.go
Comment (mx-psi) [inline exporter/datadogexporter/translate_traces.go]: You can do
```suggestion
	if len(s.ParentSpanID().Bytes()) > 0 {
		span.ParentID = decodeAPMId(hex.EncodeToString(s.ParentSpanID().Bytes()[:]))
	} 
```

File: exporter/datadogexporter/translate_traces.go
Comment (mx-psi) [inline exporter/datadogexporter/translate_traces.go]: nit: We always use `decodeAPMId` after calling `hex.EncodeToString(id.Bytes()[:])` for a certain `id`, maybe we should move that inside `decodeAPMId` too

File: exporter/datadogexporter/traces_exporter.go
Comment (ericmustin) [reply exporter/datadogexporter/traces_exporter.go]: so, looked through this again and i think we'd have issues since we're exporting potentially multiple `ddTracePayload` since we iterate over the `range aggregatedTraces`. It's one payload per env. So, tl;dr I don't think we could easily support using Otel default retries here, or at least i'm not quite how we'd handle it right now

**Review by mx-psi — COMMENTED** [review]
Looking better! I left some more comments, I think we would benefit from some unit test coverage at this point. One thing I am worried about is that the Agent might spam logs when importing its packages/running its internal functions, have you noticed anything in the end to end tests?

File: exporter/datadogexporter/traces_exporter.go
Comment (mx-psi) [reply exporter/datadogexporter/traces_exporter.go]: Okay, guess we will have to deal with it :( I still think we could benefit from having a function like

 ```go
func pushWithRetry(maxRetries int, waitTime time.Duration, fn func() error) error {
  // ...
}
```

that abstracts the retrying logic in case we want to change it afterwards, but feel free to mark this as done if you think that's not worth the effort.

Comment (ericmustin) [reply exporter/datadogexporter/traces_exporter.go]: yea i think that's reasonable, let me see if i can make this a bit cleaner

File: exporter/datadogexporter/trace_connection.go
Comment (ericmustin) [reply exporter/datadogexporter/trace_connection.go]: ah good catch, i'll update (will try to check if somethiing on the backend relies on this)

File: exporter/datadogexporter/traces_exporter.go
Comment (ericmustin) [reply exporter/datadogexporter/traces_exporter.go:85]: There's some span processors that allow regex to modify/filter specific attribs. 

- https://github.com/open-telemetry/opentelemetry-collector/tree/master/processor/attributesprocessor
- https://github.com/open-telemetry/opentelemetry-collector/tree/master/processor/spanprocessor

It's independent from a datadog exporters but i think it would be nice to eventually have documentation of suggested best practices for pipeline setups, or even plug n' play yaml examples for different use cases that we could have live in either a datadog exporter specific readme or docs section.

File: exporter/datadogexporter/translate_traces.go
Comment (ericmustin) [reply exporter/datadogexporter/translate_traces.go]: updated this to move the `hex.encodeToString` bit within the helper func. i prefer to leave it as accepting a byte array rather than a specific span data field, i think it makes testing a bit easier.

File: exporter/datadogexporter/translate_traces.go
Comment (ericmustin) [reply exporter/datadogexporter/translate_traces.go]: makes sense, updated

File: exporter/datadogexporter/traces_exporter.go
Comment (ericmustin) [reply exporter/datadogexporter/traces_exporter.go]: I tried to add at least the structure of how we'd use a retry  func here. wiring up all the wait Duration retry stuff seemed a bit complicated so i passed on that for the moment, but i think we now at least have the structure where we can just update the error handling function if we want to handle retries differently.

I also dropped the retry amount to 0 (passing in 1), since it's unclear to me if we dedupe in most cases, so would rather not experiment, if we experience lots of drops we can iterate.

**Comment by ericmustin** [comment]
@mx-psi I tried to address most of the feedback and add unit tests for the span/trace translation work. Some of the trace connection stuff is still not covered in tests, I think I have to wire up a mock server for some of that, and I wasn't totally sure how to manage merging up the config tags. But punting on those bits for now so I can work on rebasing off the most up to date metrics exporter work so that the config is in a good place. I'd also like to start gameplanning on how to start the upstream PR process if you think this is in a reasonably good place, so let me know if there are other areas here where the code needs to be fixed before that.

**Comment by ericmustin** [comment]
Added some additional tests, rebasing off master brought in a bunch of changes to metric stuff that probably is now out of date. Going to wait for your metrrics PR and try to rebase off that branch

**Comment by ericmustin** [comment]
@mx-psi So i think this is pretty much in a good state, i added mock server tests to test the trace connection so i think our code coverage is pretty good at this point. I just need to rebase off master + your metric exporter work

File: exporter/datadogexporter/model.go
Comment (mx-psi) [inline exporter/datadogexporter/model.go]:
```suggestion
		// The tracer can mark a span to be analyzed, with a value 0-1, where 1 is always keep, and 0 is always reject.
```

File: exporter/datadogexporter/trace_connection.go
Comment (mx-psi) [inline exporter/datadogexporter/trace_connection.go]: Not that it matters if we are only retrying once but (morally) shouldn't we break out of the loop if `!shouldRetry`?

File: exporter/datadogexporter/trace_connection.go
Comment (mx-psi) [inline exporter/datadogexporter/trace_connection.go]: ditto

File: exporter/datadogexporter/traces_exporter.go
Comment (mx-psi) [inline exporter/datadogexporter/traces_exporter.go]: I think this is fixed
```suggestion
```

File: exporter/datadogexporter/traces_exporter.go
Comment (mx-psi) [reply exporter/datadogexporter/traces_exporter.go:85]: Okay! I created OTEL-19 internally to keep track of this; we can mark as solved

File: exporter/datadogexporter/translate_traces.go
Comment (mx-psi) [inline exporter/datadogexporter/translate_traces.go]:
```suggestion
		// appends a specific piece of metadata to these spans marking them as analyzed
```

File: exporter/datadogexporter/translate_traces.go
Comment (mx-psi) [inline exporter/datadogexporter/translate_traces.go]: Did you find out why this was happening? A bug on the receiver/client end?

**Review by mx-psi — APPROVED** [review]
Looks good, I have yet to test it on the backend but since you already did I am going to approve. I left a bunch of small typos to fix/TODOs that can be removed too.

File: exporter/datadogexporter/trace_connection.go
Comment (ericmustin) [reply exporter/datadogexporter/trace_connection.go]: good catch, yup will update

File: exporter/datadogexporter/translate_traces.go
Comment (ericmustin) [reply exporter/datadogexporter/translate_traces.go]: i believe it's related to the jThrift span translation code with the jaeger receiver used in our e2e tests, but i haven't have a chance to create a reproduction (its possibly they python client too). I'm updating this note for future reference

**Comment by mx-psi** [comment]
Should we close this now that it is available upstream?

**Pull request closed without merging**
